### DISCUSSION TITLE
Different device selection for the same platform like GPU.

### DISCUSSION THREAD

**Discussion by yinrun** [intel/llvm#4035] [Q&A]
Hi all. I am working on a server with multiple accelerator devices such as GPUs. When I as creating several sycl queues, I want to assign each queue to the specific GPU device I assigned. In cuda, we can use `cudaSetDevice` to choose the specific device , in dpc++ runtime, the selection seems to be made in `device_selector::select_device()` and it will select a default gpu device, how can I bind different queue to different accelerators respectively?

**Comment by AlexeySachkov** [comment]
Hi @yinrun, 

> When I as creating several sycl queues, I want to assign each queue to the specific GPU device I assigned.

This is for sure possible, just take a look at constructors available for [`sycl::queue`](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#_queue_interface): you explicitly pass `device` which interest you or pass `device_selector` to choose it during queue creation.

>  in dpc++ runtime, the selection seems to be made in `device_selector::select_device()` and it will select a default gpu device

That is because you are likely using default selector - it indeed contains some hardcoded logic which favors GPUs. However, there are other selectors available according to the [spec](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:device-selector) and you can even create your own:

```c++
int prefer_my_vendor(const sycl::device & d) {
  // Return 1 if the vendor name is "MyVendor" or 0 else.
  // 0 does not prevent another device to be picked as a second choice
  return d.get_info<info::device::vendor>() == "MyVendor";
}

// Get the preferred device or another one if not available
sycl::device preferred_device { prefer_my_vendor };
```

The example is copied from the spec section I linked above.

Let me know if you other questions

**Reply by yinrun** [reply]
Thanks Alexey, the device_select is implemented in device_selector.hpp, such as cpu_selector, gpu_selector, and they override operator(), I am not sure where I can set different devices.

Maybe I didn't state my problem clearly. here is a pseudo code for example,
```
  device d1 = gpu_selector.select_device(0); // select gpu 0
  device d2 = gpu_selector.select_device(1); // select gpu 1

  queue Queue1(d1);
  queue Queue2(d2);
```
gpu 0 and gpu 1 is the same type and such kind of selector like gpu_selector.select_device(int) seems not support.

**Reply by AlexeySachkov** [reply]
> and they override operator(), I am not sure where I can set different devices.

That is correct, device selector is a [callable object](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#device-selector) which is called for every available device computing the score for it. Device with the highest score is returned to you.

Therefore, you have a few options to explore:

1. Create your own device selector in form of a functor (`struct` with `operator()`) which remembers which device was selected the first time to avoid selecting it for the second time. The idea is to select device for the first time, store selected device into device selector and select device for the second time.
2. use `device::get_devices()` to get list of devices and manually filter them selecting whichever you want
3. Another option for device selector: if you know some unique properties which are different for the first and second device, then you can just prepare two device selectors that look for those properties - it could be a vendor name, device name, driver version, etc.

**Reply by alexbatashev** [reply]
@yinrun  Alexey is probably referring to [`platform::get_devices()`](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#_platform_interface) You can do something like:

```c++
#include <algorithm>
...

sycl::platform Platform{...}; // somehow select your platform, e.g. with device selector
auto Devices = Platform.get_devices();
std::vector<sycl::queue> Queues;
std::transform(Devices.begin(), Devices.end(), std::back_inserter(Queues), [](const sycl::device &D) {
  return sycl::queue{D};
});
```

This will give you one queue per each device in the platform. 

Also there's a [FilterSelector](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/FilterSelector/FilterSelector.adoc).

**Reply by yinrun** [reply]
Hi @alexbatashev @AlexeySachkov , `Platform.get_devices() ` is a very good idea. Howerver, in my situation, I have two seperate progress to use different devices and all the devices are totally the same(3rd option maybe not suitable).  for the 1st option, do you mean I need something like a global id in method such as `int gpu_selector::operator()(const device &dev) const', this method is used in `select_device()` method and I don't know how to pass a parameter like device_id to it.

**Reply by alexbatashev** [reply]
If you somehow can identify your processes, you can go for filter_selector, which accepts strings in format `cuda:gpu:0`. Just replace 0 with gpu id. It is guaranteed to be stable across runs (but not across different machines, obviously).

**Reply by yinrun** [reply]
Thanks @alexbatashev I think this is exactly what I want. I will try it.
